runtime/compiler/optimizer/InterpreterEmulator.cpp

```diff
@@ -1023,6 +1023,44 @@ InterpreterEmulator::getReturnValue(TR_ResolvedMethod *callee)
             }
          break;
          }
+      case TR::java_lang_invoke_VarHandle_asDirect:
+      case TR::java_lang_invoke_Invokers_directVarHandleTarget:
+         {
+         Operand* varHandleOperand = top();
+         TR::KnownObjectTable::Index vhIndex = varHandleOperand->getKnownObjectIndex();
+         TR::KnownObjectTable *knot = comp()->getKnownObjectTable();
+         if (knot
+              && vhIndex != TR::KnownObjectTable::UNKNOWN
+              && !knot->isNull(vhIndex))
+            {
+            const char * varHandleClassName = "java/lang/invoke/VarHandle";
+            TR_OpaqueClassBlock * varHandleClass =
+               comp()->fej9()->getSystemClassFromClassName(varHandleClassName, strlen(varHandleClassName));
+            TR_OpaqueClassBlock * objectClass = comp()->fej9()->getObjectClassFromKnownObjectIndex(comp(), vhIndex);
+            if (fe()->isInstanceOf(objectClass, varHandleClass, true) == TR_yes)
```

Comment:
I don't think this type test does what we need. If the handle is an instance of `IndirectVarHandle`, then `isInstanceOf()` will still be `TR_yes`, since that's a subtype of `VarHandle`

There are a few ways to go here:
1. We could check for an exact type match, i.e. `objectClass == varHandleClass`. This would also exclude any other subclasses of `VarHandle`
2. We could check to make sure it isn't an instance of `IndirectVarHandle` instead. This would work for other subclasses, but it could fail in the future if a new subclass overrides the default implementation
3. We could check the vtable to see whether `objectClass` has inherited the default implementation. This would be the most precise, but it would also be the most work

I'm not sure which way is the best at the moment, but I'm inclined to avoid approach 2 because it's the most brittle. However, it might be worth noting that there's some inherent brittleness just due to the use of recognized methods. That is, a later change to `VarHandle.asDirect()` or `directVarHandleTarget()` could break this logic even with approach 1, which is the safest of the three

Reply:
I decided to go with option 1.